File: algorithms/easy/delete-node-in-a-linked-list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def deleteNode(self, node):
        """
        :type node: ListNode
        :rtype: void Do not return anything, modify node in-place instead.
        """
        while node and node.next:
            node.val = node.next.val
            if node.next.next is None:
                node.next = None
            node = node.next


def test():
    head = ListNode(0)
    n1 = ListNode(1)
    n2 = ListNode(2)
    n3 = ListNode(3)
    n4 = ListNode(4)
    n5 = ListNode(5)
    head.next = n1
    n1.next = n2
    n2.next = n3
    n3.next = n4
    n4.next = n5
    s = Solution()
    s.deleteNode(n2)
    l = []
    while head:
        l.append(head.val)
        head = head.next
    assert l == [0, 1, 3, 4, 5]

if __name__ == '__main__':
    test()

File: algorithms/easy/delete-node-in-a-linked-list.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};
struct ListNode *head;

struct ListNode* newNode(int val, struct ListNode *next) {
    struct ListNode *node = malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = next;
    return node;
}

void deleteNode(struct ListNode* node) {
    if (node == NULL)
        return;
    node->val = node->next->val;
    node->next = node->next->next;
}

int main() {
    head = newNode(1, newNode(2, newNode(3, newNode(4, newNode(5, NULL)))));
    deleteNode(head);
    printf("\n");
    while(head != NULL) {
        printf("%d ", head->val);
        head = head->next;
    }
    printf("\n");

    struct ListNode *a = newNode(2, newNode(3, newNode(2, newNode(5, NULL))));
    head = newNode(1, a);
    deleteNode(a);
    printf("\n");
    while(head != NULL) {
        printf("%d ", head->val);
        head = head->next;
    }
    printf("\n");
}
